plugins/hls-qualify-imported-names-plugin/src/Ide/Plugin/QualifyImportedNames.hs

```diff
@@ -0,0 +1,259 @@
+
+instance Ord IdentifierSpan where
+  compare (IdentifierSpan line1 startCol1 endCol1) (IdentifierSpan line2 startCol2 endCol2) =
+    (line1 `compare` line2) `thenCmp` (startCol1 `compare` startCol2) `thenCmp` (endCol1 `compare` endCol2)
+
+realSrcSpanToIdentifierSpan :: Span -> Maybe IdentifierSpan
+realSrcSpanToIdentifierSpan realSrcSpan
+  | let startLine = srcSpanStartLine realSrcSpan - 1
+  , let endLine = srcSpanEndLine realSrcSpan - 1
+  , startLine == endLine
+  , let startCol = srcSpanStartCol realSrcSpan - 1
+  , let endCol = srcSpanEndCol realSrcSpan - 1 =
+      Just $ IdentifierSpan startLine startCol endCol
+  | otherwise = Nothing
+
+identifierSpanToRange :: IdentifierSpan -> Range
+identifierSpanToRange (IdentifierSpan line startCol endCol) =
+  Range (Position line startCol) (Position line endCol)
+
+data UsedIdentifier = UsedIdentifier {
+  usedIdentifierName :: !Name,
+  usedIdentifierSpan :: !IdentifierSpan
+}
+
+refMapToUsedIdentifiers :: RefMap a -> [UsedIdentifier]
+refMapToUsedIdentifiers = DList.toList . Map.foldlWithKey' folder DList.empty
+  where
+    folder acc identifier spanIdentifierDetailsPairs =
+      DList.fromList (mapMaybe (uncurry (getUsedIdentifier identifier)) spanIdentifierDetailsPairs) <> acc
+
+    getUsedIdentifier :: Identifier -> Span -> IdentifierDetails a -> Maybe UsedIdentifier
+    getUsedIdentifier identifier span IdentifierDetails {..}
+      | Just identifierSpan <- realSrcSpanToIdentifierSpan span
+      , Right name <- identifier
+      , Use `elem` identInfo = Just $ UsedIdentifier name identifierSpan
+      | otherwise = Nothing
+
+occNameToText :: OccName -> Text
+occNameToText = Text.pack . occNameString
+
+updateColOffset :: Int -> Int -> Int -> Int
+updateColOffset row lineOffset colOffset
+  | row == lineOffset = colOffset
+  | otherwise = 0
+
+usedIdentifiersToTextEdits :: Range -> NameEnv [ImportedBy] -> Text -> [UsedIdentifier] -> [TextEdit]
+usedIdentifiersToTextEdits range nameToImportedByMap sourceText usedIdentifiers
+  | let sortedUsedIdentifiers = sortOn usedIdentifierSpan usedIdentifiers =
+      State.evalState (makeStateComputation sortedUsedIdentifiers) (Text.lines sourceText, 0, 0)
+  where
+    folder :: [TextEdit] -> UsedIdentifier -> State ([Text], Int, Int) [TextEdit]
+    folder prevTextEdits (UsedIdentifier identifierName identifierSpan)
+      | Just importedBys <- lookupNameEnv nameToImportedByMap identifierName
+      , Just (ImportedBy alias _) <- find (isRangeWithinImportedBy range) importedBys
+      , let IdentifierSpan row startCol endCol = identifierSpan
+      , let identifierRange = identifierSpanToRange identifierSpan
+      , let aliasText = Text.pack $ moduleNameString alias
+      , let identifierText = Text.pack $ occNameString $ nameOccName identifierName
+      , let qualifiedIdentifierText = aliasText <> "." <> identifierText = do
+          (sourceTextLines, lineOffset, updateColOffset row lineOffset -> colOffset) <- State.get
```

Comment:
I suppose that this is to account for multiple identifiers in the same line. 
Is there a test for that?

Reply:
Yes. I didn't know a way to see whether a name is parenthesized, or backticked in the parsing pass, and saw an example in the "suggest import disambiguation" code action where they simply checked the span in the source. So I did the same thing except more involved because I wanted one pass through the file for all names.

I added a test to show parenthesized and backticked names on the same line get qualified. For just regular names on the same line there are other tests that show this.